### PR TITLE
aad connet timeout

### PR DESCRIPTION
**Pull request by sugus-seu**
[no connect timeout #106](https://github.com/ninenines/gun/issues/106)

### CODE REVIEW

**Comment by essen** [comment]
Use Opts, don't add an argument.

**Comment by sugus-seu** [comment]
I use an argument because ranch did that way, the opts are more like options of net layer

**Comment by sugus-seu** [comment]
I use an argument because ranch did that way, the opts are more like options of net layer

**Comment by sugus-seu** [comment]
I use an argument because ranch did that way, the opts are more like options of net layer

**Comment by sugus-seu** [comment]
I use an argument because ranch did that way, the opts are more like options of net layer

**Comment by sugus-seu** [comment]
I use an argument because ranch did that way, the opts are more like options of net layer

**Comment by essen** [comment]
The options are not transport layer options. Use Opts.

**Comment by sugus-seu** [comment]
done

File: src/gun.erl [hunk 1]
```diff
@@ -473,7 +475,7 @@ dbg_send_raw(ServerPid, Data) ->
 
 start_link(Owner, Host, Port, Opts) ->
 	proc_lib:start_link(?MODULE, init,
-		[self(), Owner, Host, Port, Opts]).
+        [self(), Owner, Host, Port, Opts], maps:get(conn_timeout, Opts, 1000)).
```
Comment (essen) [inline src/gun.erl]: Don't really understand this part. We init_ack right at the start so this doesn't actually do anything. The start_link returns before we even try to connect.

File: src/gun.erl [hunk 2]
```diff
@@ -497,7 +499,7 @@ connect(State=#state{host=Host, port=Port, opts=Opts, transport=Transport=ranch_
 		{alpn_advertised_protocols, Protocols},
 		{client_preferred_next_protocols, {client, Protocols, <<"http/1.1">>}}
 		|maps:get(transport_opts, Opts, [])],
-	case Transport:connect(Host, Port, TransportOpts) of
+    case Transport:connect(Host, Port, TransportOpts, maps:get(conn_timeout, Opts, 1000)) of
```
Comment (essen) [inline src/gun.erl]: Default of 1000 is too small. 5000 is usually good.

File: src/gun.erl [hunk 3]
```diff
@@ -162,9 +162,11 @@ check_options([Opt = {protocols, L}|Opts]) when is_list(L) ->
 		_ ->
 			{error, {options, Opt}}
 	end;
+check_options([{conn_timeout, T}|Opts]) when is_integer(T), T > 0 ->
+    check_options(Opts);
 check_options([{retry, R}|Opts]) when is_integer(R), R >= 0 ->
 	check_options(Opts);
-check_options([{retry_timeout, T}|Opts]) when is_integer(T) > 0 ->
+check_options([{retry_timeout, T}|Opts]) when is_integer(T), T > 0 ->
```
Comment (essen) [inline src/gun.erl]: This is unrelated and should be in its own commit.

**Comment by sugus-seu** [comment]
commit again

**Comment by essen** [comment]
Merged with some changes, tests and docs; the option will be named `connect_timeout` and default to infinity. Thanks!
